notebook/static/notebook/js/notebook.js

```diff
@@ -2845,6 +2845,57 @@ define([
         }
     };
     
+       
+    Notebook.prototype.save_notebook_as = function() {
+            // If current notebook has some changes, save them
+            if (this.writable && this.dirty) {
+                this.save_notebook();
+            }
+            var that = this;
+            var d = that.save_widget._nb_name_dialog({
+                dialog_title: i18n.msg._('Save As'),
+                message: i18n.msg._('Enter a notebook path relative to notebook dir'),
+                message_classname: 'save-message',
+                button_name: 'Save',
+                click_handler: function(d) {
+                    var nb_path = d.find('input').val();
+                    var nb_name = nb_path.split('/').slice(-1).pop();
+                    var ext = utils.splitext(nb_name)[1];
+                    // If notebook name does not contain extension '.ipynb' add it
+                    if (ext === '') {
+                        nb_name = nb_name + '.ipynb';
+                        nb_path = nb_path + '.ipynb';
+                    }
+                    var save_thunk = function() {
+                      var model = {'type': 'notebook','content': that.toJSON(),
+                                   'name': nb_name};
+                      return that.contents.save(nb_path, model)
+                          .then(function(data) {
+                              var dir_path = utils.get_body_data("baseUrl");
+                              window.open(utils.url_path_join(dir_path, "notebooks", data.path), '_self');
```

Comment:
There's already code to deal with updating the address bar and other details when the notebook is renamed, based on the `notebook_renamed.Notebook` event. Instead of calling window.open(), we should make sure that event fires, as on this line:

https://github.com/jupyter/notebook/blob/5fdf8c23061607824cfab19f6840de692a916d1a/notebook/static/notebook/js/notebook.js#L3036

We should also ensure that the session is updated - that links a kernel to a filename. See the line above the one I just linked to.

Reply:
I'm not sure if triggering event will cause the new notebook to open up...I'll check and get back.